src/java/org/apache/cassandra/io/tries/TrieNode.java

```diff
@@ -0,0 +1,968 @@
+        {
+            assert payloadBits == 0;
+            int childCount = node.childCount();
+            assert childCount == 1;
+            long pd = -node.serializedPositionDelta(0, nodePosition);
+            assert pd > 0 && pd < 0x10;
+            dest.writeByte((ordinal << 4) + (int) (pd & 0x0F));
+            dest.writeByte(node.transition(0));
+        }
+
+        @Override
+        public int sizeofNode(SerializationNode<?> node)
+        {
+            return 2;
+        }
+    };
+
+    static final Single SINGLE_NOPAYLOAD_12 = new SingleNoPayload12();
+    static class SingleNoPayload12 extends Single
+    {
+        // 4-bit type ordinal
+        // 12-bit target delta
+        // byte transition
+        // no payload!
+        SingleNoPayload12()
+        {
+            super(FRACTIONAL_BYTES);
+        }
+
+        @Override
+        public int payloadFlags(ByteBuffer src, int position)
+        {
+            return 0;
+        }
+
+        // Although we don't have a payload position, provide one for calculating the size of the node.
+        @Override
+        public int payloadPosition(ByteBuffer src, int position)
+        {
+            return position + 3;
+        }
+
+        @Override
+        public int search(ByteBuffer src, int position, int transitionByte)
+        {
+            int c = src.get(position + 2) & 0xFF;
+            if (transitionByte == c)
+                return 0;
+            return transitionByte < c ? -1 : -2;
+        }
+
+        @Override
+        public long transitionDelta(ByteBuffer src, int position, int childIndex)
+        {
+            return -(src.getShort(position) & 0xFFF);
+        }
+
+        @Override
+        public int transitionByte(ByteBuffer src, int position, int childIndex)
+        {
+            return childIndex == 0 ? src.get(position + 2) & 0xFF : Integer.MAX_VALUE;
+        }
+
+        @Override
+        boolean fits(long delta)
+        {
+            return 0 <= delta && delta <= 0xFFF;
+        }
+
+        @Override
+        public void serialize(DataOutput dest, SerializationNode<?> node, int payloadBits, long nodePosition) throws IOException
+        {
+            assert payloadBits == 0;
+            int childCount = node.childCount();
+            assert childCount == 1;
+            long pd = -node.serializedPositionDelta(0, nodePosition);
+            assert pd > 0 && pd < 0x1000;
+            dest.writeByte((ordinal << 4) + (int) ((pd >> 8) & 0x0F));
+            dest.writeByte((byte) pd);
+            dest.writeByte(node.transition(0));
+        }
+
+        @Override
+        public int sizeofNode(SerializationNode<?> node)
+        {
+            return 3;
+        }
+    };
+
+    static final Sparse SPARSE_8 = new Sparse(1);
+    static final Sparse SPARSE_16 = new Sparse(2);
+    static final Sparse SPARSE_24 = new Sparse(3);
+    static final Sparse SPARSE_40 = new Sparse(5);
```

Comment:
nit: Should probably move these into `Sparse` itself to avoid potential classloading deadlocks